torch/distributed/distributed_c10d.py

```diff
@@ -753,10 +755,35 @@ def recv(tensor,
         return src
 
 
+def async_collective(func):
+    @functools.wraps(func)
+    def wrapped(*args, async_op=False, **kwargs):
+        work = func(*args, **kwargs)
+        if work is None:
+            return
+
+        should_profile = torch.autograd._profiler_enabled()
+        if should_profile:
+            rf = torch.autograd.profiler.record_function(func.__name__)
+        else:
+            rf = contextlib.suppress()
+
+        with rf:
+            if async_op:
+                if should_profile:
+                    fut = rf._call_end_callbacks_on_future(work.get_future())
```

Comment:
`get_future` currently only works for NCCL, I'm wondering if there is away to ensure `with record_function` works for async cuda events.